tensorflow/compiler/xla/service/gpu/horizontal_fusion.cc

```diff
@@ -215,17 +227,25 @@ void HorizontalFusionImpl::FusionCandidates::Initialize(
     }
   }
 
-  // Sort according to the number of outputs and instruction counts, because
-  // we fuse only instructions with the same number of outputs and whose
-  // computations have the same instruction counts.
-  std::sort(fusion_instrs_.begin(), fusion_instrs_.end(),
-            [&](const HloInstruction* a, const HloInstruction* b) {
-              if (GetOutputSizeOfFusion(*a) == GetOutputSizeOfFusion(*b)) {
-                return a->fused_instruction_count() <
-                       b->fused_instruction_count();
-              }
-              return GetOutputSizeOfFusion(*a) < GetOutputSizeOfFusion(*b);
-            });
+  // Sort `fusion_instrs` according to output types, the number of outputs,
+  // and instruction counts, because we only fuse instructions with the same
+  // number/type of outputs and whose computations have the same instruction
+  // count.
+  std::sort(
+      fusion_instrs_.begin(), fusion_instrs_.end(),
+      [&](const HloInstruction* a, const HloInstruction* b) {
+        PrimitiveType output_type_a =
+            GetOutputsOfFusion(*a)[0]->shape().element_type();
```

Comment:
The `[0]` is safe because `GetOutputsOfFusion` above will evaluate to `false` when it isn't. However, this is subtle and easy to break as code evolves. In fact, I looked at a very similar bug just today. Please explicitly check the assumption that fusions have a unique output type. You can introduce a helper `GetUniqueOutputType` that CHECK fails otherwise.

Reply:
Fixed. It is indeed better and clearer to have `GetUniqueOutputTypeOfFusion`. Thanks for the comments.